### PR TITLE
Introduce citus_internal_update_relation_colocation

### PR DESCRIPTION
**Pull request by onderkalaci**
update_distributed_table_colocation can be called by the relation
owner, and internally it updates pg_dist_partition. With this
commit, update_distributed_table_colocation uses an internal
UDF to access pg_dist_partition.

As a result, this operation can now be done by regular users
on MX.

### CODE REVIEW

File: src/test/regress/sql/update_colocation_mx.sql [hunk 1]
```diff
@@ -0,0 +1,45 @@
+
+-- in order to make the enterprise and community
+-- tests outputs the same, disable enable_ddl_propagation
+-- and create the roles/schema manually
```
Comment (onderkalaci) [inline src/test/regress/sql/update_colocation_mx.sql:4]: verified that this test passes on enterprise

File: src/backend/distributed/metadata/metadata_sync.c [hunk 2]
```diff
@@ -2612,3 +2612,63 @@ citus_internal_delete_shard_metadata(PG_FUNCTION_ARGS)
...
+			/* connection from the coordinator operating on a shard */
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("Updating colocation ids are only allowed for hash "
+								   "distributed tables: %c", partitionMethod)));
+		}
+
+		int count = 1;
+		List *targetColocatedTableList =
+			ColocationGroupTableList(tagetColocationId, count);
+		if (list_length(targetColocatedTableList) == 0)
```
Comment (JelteF) [inline src/backend/distributed/metadata/metadata_sync.c:2656]: If we want to break colocation, we should allow this right?

File: src/backend/distributed/sql/udfs/citus_internal_update_relation_colocation/latest.sql [hunk 3]
```diff
@@ -0,0 +1,7 @@
+CREATE OR REPLACE FUNCTION pg_catalog.citus_internal_update_relation_colocation(relation_id Oid, target_colocation_id int)
+    RETURNS void
+    LANGUAGE C STRICT
+    AS 'MODULE_PATHNAME';
+COMMENT ON FUNCTION pg_catalog.citus_internal_update_relation_colocation(oid, int) IS
+    'Deletes rows from pg_dist_shard and pg_dist_shard_placement with user checks';
```
Comment (JelteF) [inline src/backend/distributed/sql/udfs/citus_internal_update_relation_colocation/latest.sql:6]: comment is incorrect

File: src/backend/distributed/utils/colocation_utils.c [hunk 4]
```diff
@@ -887,9 +892,14 @@ ColocatedTableList(Oid distributedTableId)
...
+ *
+ * If count is zero then the command is executed for all rows that it applies to.
+ * If count is greater than zero, then no more than count rows will be retrieved;
+ * execution stops when the count is reached, much like adding a LIMIT clause
+ * to the query.
  */
-static List *
-ColocationGroupTableList(Oid colocationId)
+List *
+ColocationGroupTableList(uint32 colocationId, bool count)
```
Comment (JelteF) [inline src/backend/distributed/utils/colocation_utils.c:902]:
```suggestion
ColocationGroupTableList(uint32 colocationId, int32 count)
```

File: src/test/regress/sql/update_colocation_mx.sql [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+SELECT update_distributed_table_colocation('t1', 't2');
+
+-- show that we successfuly updated the colocationids to the same value
+SELECT count(DISTINCT colocationid) FROM pg_dist_partition WHERE logicalrelid IN ('t1'::regclass, 't2'::regclass);
+\c - - - :worker_1_port
+SET search_path TO "Update Colocation";
+SELECT count(DISTINCT colocationid) FROM pg_dist_partition WHERE logicalrelid IN ('t1'::regclass, 't2'::regclass);
+\c - - - :worker_2_port
+SET search_path TO "Update Colocation";
+SELECT count(DISTINCT colocationid) FROM pg_dist_partition WHERE logicalrelid IN ('t1'::regclass, 't2'::regclass);
```
Comment (JelteF) [inline src/test/regress/sql/update_colocation_mx.sql:41]: Needs a test for `update_distributed_table_colocation('t1', 'none');` To show we can break colocation in MX too.

File: src/backend/distributed/metadata/metadata_sync.c (hunk 2 above)
Comment (onderkalaci) [reply src/backend/distributed/metadata/metadata_sync.c:2656]: ops, I didn't realize that we can break colocation, let me check what is needed here

File: src/backend/distributed/metadata/metadata_sync.c [hunk 6]
```diff
@@ -2612,3 +2612,63 @@ citus_internal_delete_shard_metadata(PG_FUNCTION_ARGS)
...
+		{
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("No table belongs to target colocation id: %d",
+								   tagetColocationId)));
+		}
+
+		Oid targetRelationId = linitial_oid(targetColocatedTableList);
+
+		ErrorIfShardPlacementsNotColocated(relationId, targetRelationId);
```
Comment (JelteF) [inline src/backend/distributed/metadata/metadata_sync.c:2665]: I'm not sure if this check is really needed. If a malicious user wants to have a table that is marked colocated, but isn't actually colocated. Then they can easily do so by creating a new table and not matching up the shards.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 7]
```diff
@@ -2612,3 +2612,63 @@ citus_internal_delete_shard_metadata(PG_FUNCTION_ARGS)
...
+		CheckReplicationModel(relationId, targetRelationId);
+		CheckDistributionColumnType(relationId, targetRelationId);
```
Comment (JelteF) [inline src/backend/distributed/metadata/metadata_sync.c:2667]: These checks should also be part of `citus_internal_add_partition_metadata`.

**Review by JelteF — CHANGES_REQUESTED** [review]
Two errors are not covered in tests according to codecov.

File: src/backend/distributed/metadata/metadata_sync.c (hunk 6 above)
Comment (onderkalaci) [reply src/backend/distributed/metadata/metadata_sync.c:2665]: That's true, but why not prevent it as much as possible? This function would be used very infrequently, and I think we should be OK with the overhead of this check.

File: src/backend/distributed/metadata/metadata_sync.c (hunk 2 above)
Comment (onderkalaci) [reply src/backend/distributed/metadata/metadata_sync.c:2707]: now removed this check and added a test

File: src/backend/distributed/sql/udfs/citus_internal_update_relation_colocation/latest.sql (hunk 3 above)
Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/citus_internal_update_relation_colocation/latest.sql]: updated

File: src/backend/distributed/utils/colocation_utils.c (hunk 4 above)
Comment (onderkalaci) [reply src/backend/distributed/utils/colocation_utils.c]: done

File: src/test/regress/sql/update_colocation_mx.sql (hunk 5 above)
Comment (onderkalaci) [reply src/test/regress/sql/update_colocation_mx.sql:41]: added now

File: src/backend/distributed/utils/colocation_utils.c [hunk 8]
```diff
@@ -728,12 +719,47 @@ CheckDistributionColumnType(Oid sourceRelationId, Oid targetRelationId)
...
 
 
+/*
+ * GetColumnTypeEquality checks if distribution column types and collations
+ * of the given columns are same. The function sets the boolean pointers.
+ */
+void
+GetColumnTypeEquality(Var *sourceDistributionColumn, Var *targetDistributionColumn,
+					  bool *columnTypesSame, bool *columnCollationsSame)
+{
```
Comment (onderkalaci) [inline src/backend/distributed/utils/colocation_utils.c:729]: this part is just a refactor of `CheckDistributionColumnType ` as we only need `GetColumnTypeEquality` in  `citus_internal_add_partition_metadata`

File: src/backend/distributed/metadata/metadata_sync.c [hunk 9]
```diff
@@ -2155,6 +2156,55 @@ EnsurePartitionMetadataIsSane(char distributionMethod, int colocationId,
...
+	{
+		int count = 1;
+		List *targetColocatedTableList =
+			ColocationGroupTableList(colocationId, count);
+
+		/*
+		 * If we have any colocated hash tables, ensure if they share the
+		 * same distribution key properties.
+		 */
+		if (list_length(targetColocatedTableList) == 1)
```
Comment (JelteF) [inline src/backend/distributed/metadata/metadata_sync.c:2170]:
```suggestion
		if (list_length(targetColocatedTableList) >= 1)
```

File: src/backend/distributed/utils/colocation_utils.c [hunk 10]
```diff
@@ -728,12 +719,47 @@ CheckDistributionColumnType(Oid sourceRelationId, Oid targetRelationId)
 }
 
 
+/*
+ * GetColumnTypeEquality checks if distribution column types and collations
+ * of the given columns are same. The function sets the boolean pointers.
+ */
+void
+GetColumnTypeEquality(Var *sourceDistributionColumn, Var *targetDistributionColumn,
```
Comment (JelteF) [inline src/backend/distributed/utils/colocation_utils.c:727]: I think we can change this to `CheckColumnTypeEquality`, then we can include the error messages in here as well so we don't need to duplicate them.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 11]
```diff
@@ -2612,3 +2662,64 @@ citus_internal_delete_shard_metadata(PG_FUNCTION_ARGS)
...
+								   "entry in pg_dist_partition.",
+								   get_rel_name(relationId))));
+		}
+		else if (partitionMethod != DISTRIBUTE_BY_HASH)
+		{
+			/* connection from the coordinator operating on a shard */
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("Updating colocation ids are only allowed for hash "
+								   "distributed tables: %c", partitionMethod)));
+		}
```
Comment (JelteF) [inline src/backend/distributed/metadata/metadata_sync.c:2701]: These two checks don't have coverage

**Review by JelteF — APPROVED** [review]
Some final things, but after that this is good.

File: src/backend/distributed/metadata/metadata_sync.c (hunk 9 above)
Comment (onderkalaci) [reply src/backend/distributed/metadata/metadata_sync.c:2170]: We pass `count = 1` to  `ColocationGroupTableList`, that's why I used `== 1`. But, yes, lets not have such a dependency. I'll update.

(ps: I don't like committing the suggestion to remote branches as it complicates my git workflow. Instead, I manually locally)

Comment (JelteF) [reply src/backend/distributed/metadata/metadata_sync.c:2170]: > (ps: I don't like committing the suggestion to remote branches as it complicates my git workflow. Instead, I manually locally)

Typing `git pull` is an issue?

File: src/backend/distributed/utils/colocation_utils.c (hunk 10 above)
Comment (onderkalaci) [reply src/backend/distributed/utils/colocation_utils.c:727]: makes sense

File: src/backend/distributed/metadata/metadata_sync.c (hunk 9 above)
Comment (onderkalaci) [reply src/backend/distributed/metadata/metadata_sync.c:2170]: Yes, I'm not comfortable with `git pull` for my own branches :)

File: src/backend/distributed/metadata/metadata_sync.c (hunk 11 above)
Comment (onderkalaci) [reply src/backend/distributed/metadata/metadata_sync.c:2672]: manually modified metatada in a tx block to trigger
